spatialprofilingtoolbox/ondemand/providers/squidpy_provider.py

```diff
@@ -0,0 +1,107 @@
+                    ON sc.component_study=fsn.study
+                JOIN study_component sc2
+                    ON sc2.primary_study=sc.primary_study
+            WHERE sc2.component_study=%s AND
+                (
+                    (fs.specifier=%s AND fs.ordinality='1')'''
+        for i, phenotype in enumerate(phenotypes):
+            if i > 0:
+                query += f'''
+                    OR (fs.specifier=%s AND fs.ordinality={i+1})'''
+            args.append(phenotype)
+        query += '''
+                ) AND
+                fsn.derivation_method=%s
+        ;
+        '''
+        args.append(describe_squidpy_feature_derivation_method())
+        with DBCursor() as cursor:
+            cursor.execute(query, args)
+            rows = cursor.fetchall()
+        feature_specifications = {row[0]: [] for row in rows}
+        for row in rows:
+            feature_specifications[row[0]].append(row[1])
+        for key, specifiers in feature_specifications.items():
+            if len(specifiers) == len(phenotypes):
+                return key
+        message = 'Creating feature with specifiers: (%s) %s'
+        for _ in range(len(phenotypes)-1):
+            message += ', %s'
+        logger.debug(message, study_name, *phenotypes)
+        return SquidpyProvider._create_feature_specification(study_name, phenotypes)
+
+    @staticmethod
+    def _create_feature_specification(study_name: str, phenotypes: list[str]) -> str:
+        # This is close enough to the ProximityProvider implementation that it can be abstracted
+        # into PendingProvider, but it's short enough that I don't want to bother.
+        method = describe_squidpy_feature_derivation_method()
+        with DBCursor() as cursor:
+            Uploader = ADIFeatureSpecificationUploader
+            feature_specification = Uploader.add_new_feature(
+                phenotypes, method, study_name, cursor)
+        return feature_specification
+
+    def have_feature_computed(self, feature_specification: str) -> None:
+        # Ditto for this one.
+        study_name, specifiers = SquidpyProvider.retrieve_specifiers(
+            feature_specification)
+        phenotypes: list[PhenotypeCriteria] = [phenotype_str_to_phenotype(s) for s in specifiers]
+        sample_identifiers = SquidpyProvider.get_sample_identifiers(feature_specification)
+        for sample_identifier in sample_identifiers:
+            values = compute_squidpy_metrics_for_one_sample(
+                self.get_cells(sample_identifier, study_name), phenotypes)
+            # TODO: Figure out how to upload vector and matrix squidpy metrics.
+            message = 'Computed feature values of %s: %s, %s'
+            logger.debug(message, feature_specification, sample_identifier, values)
+            with DBCursor() as cursor:
+                for value in values:
+                    add_feature_value(feature_specification,
+                                      sample_identifier, value, cursor)
```

Comment:
> I don't see a reason not to use `feature_specification` as it was designed, for each feature. If we compute 5 features using the Squidpy library, they can just be documented in that system (feature specification with specifiers etc.). The apparently complicated type expression is of course not that complicated when understood as the disjunction of 5 relatively unrelated types. Why do you think we should not do this?

It's not 5 features, it's more on like 5,000 because it's a matrix. For each sample. And it's not consistent either, it's dependent on how many clusters you choose to have when you input phenotypes. 2 for 1 phenotype, 3 for 2 phenotypes, and so on.

If my understanding is correct, that's thousands of features per sample, each with their own feature specification, from a single study-clustering setup. That seems unwieldy to me with the current setup.

We should either find a new way of storing features, change the current way to accommodate vectors and matrices, or figure out how to boil down the large mass of squidpy return values into one or a few floats.

Reply:
I don't think this counts as a feature at all. I would call a 5000-entry matrix computed from source data the result of a data transformation. "Feature" connotes quantitative summary which has a human-understandable interpretation. From the `scstudies` schema:

**[Feature derivation method](https://adiframework.com/docs_site/scstudies_entities_alphabetical.html#feature_derivation_method)**. *The plan for a computational or experimental process that results in associations of quantitative (or, sometimes, qualitative) values to elements of a sample set. Often the associations involve an aggregation over finer-grained samples to a coarsest level of granularity, and are dependent on enumerated specifiers, like an ordered tuple of genes or image channels.*

Reply:
I will take a stab at writing the descriptions of the feature derivation method for the discernible outputs of squidpy.